File: DSA/src/Questions/Array/MaxConsecutiveOnesIII.java
package Questions.Array;
/*
https://leetcode.com/problems/max-consecutive-ones-iii/description/
Given a binary array nums and an integer k, return the maximum number of consecutive 1's in the array if you can flip at most k 0's.

 

Example 1:

Input: nums = [1,1,1,0,0,0,1,1,1,1,0], k = 2
Output: 6
Explanation: [1,1,1,0,0,1,1,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.
Example 2:

Input: nums = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1], k = 3
Output: 10
Explanation: [0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.
 

Constraints:

1 <= nums.length <= 105
nums[i] is either 0 or 1.
0 <= k <= nums.length
Explanation:
Initialization:

left and right pointers are initialized at the start of the array.
zeroCount keeps track of the number of zeros in the current window.
maxLen stores the maximum length of the subarray with at most k zeros.
Window Expansion:

Iterate through the array using the right pointer. If a zero is encountered, increment zeroCount.
Window Shrinking:

If zeroCount exceeds k, increment the left pointer until zeroCount is less than or equal to k.
Update Maximum Length:

For each valid window (where zeroCount ≤ k), update maxLen with the current window size (right - left + 1).
Time Complexity:
O(n): We only traverse the array once with the right pointer, and each element is processed at most twice (once by right and once by left).
Space Complexity:
O(1): We use only a few integer variables for counting and pointer manipulation, regardless of the input size.
 */
public class MaxConsecutiveOnesIII {
    public int longestOnes(int[] nums, int k) {
        int left = 0, right = 0;
        int maxLen = 0;
        int zeroCount = 0;

        while (right < nums.length) {
            if (nums[right] == 0) {
                zeroCount++;
            }

            // If the number of zeros exceeds k, shrink the window from the left
            while (zeroCount > k) {
                if (nums[left] == 0) {
                    zeroCount--;
                }
                left++;
            }

            // Calculate the maximum length of the window
            maxLen = Math.max(maxLen, right - left + 1);

            right++;
        }

        return maxLen;
    }
    public static void main(String[] args) {
        MaxConsecutiveOnesIII maxConsecutiveOnesIII = new MaxConsecutiveOnesIII();
        int[] nums = {1,1,1,0,0,0,1,1,1,1,0};
        int k = 2;
        System.out.println(maxConsecutiveOnesIII.longestOnes(nums, k)); // 6
    }
}

File: DSA/Python/Questions/Array/MaxConsecutiveOnesIII.py
"""
# Question: Max Consecutive Ones III
# Link: https://leetcode.com/problems/max-consecutive-ones-iii/

# Given a binary array nums and an integer k, return the maximum number of consecutive 1's 
# in the array if you can flip at most k 0's.

# Time Complexity: O(n) where n is the length of input array
# Space Complexity: O(1) using sliding window approach

# Algorithm:
# 1. Use sliding window technique with left and right pointers
# 2. Expand window while counting zeros
# 3. When zero count exceeds k, shrink window from left
# 4. Track maximum window length throughout process

# Key Components:
# - longest_ones(): Implements sliding window approach
# - Zero count tracking within window
# - Dynamic window size adjustment
"""

from typing import List


class MaxConsecutiveOnesIII:
    def longest_ones(self, nums: List[int], k: int) -> int:
        left = right = 0
        max_len = 0
        zero_count = 0

        while right < len(nums):
            if nums[right] == 0:
                zero_count += 1

            # Shrink window if zero count exceeds k
            while zero_count > k:
                if nums[left] == 0:
                    zero_count -= 1
                left += 1

            # Update maximum length
            max_len = max(max_len, right - left + 1)
            right += 1

        return max_len


def main():
    solution = MaxConsecutiveOnesIII()
    nums = [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0]
    k = 2
    result = solution.longest_ones(nums, k)
    print(result)  # Expected output: 6


if __name__ == "__main__":
    main()
